### ISSUE TITLE
Incorrect results with collision objects of type Convex

### ISSUE THREAD

**Issue by JeffHough** [humanoid-path-planner/hpp-fcl#346]
Hello! I have been moving over my collision checks to use collision objects of type `Convex` instead of `BVHModel`, but the results I am getting are incorrect. I think I may be missing something in the setup or usage of these objects, but I am not sure.

As explained in #340, I am doing collision checks on non-convex objects using groups of convex sub-meshes. Each of those sub-meshes is now considered its own `CollisionObject` of type `Convex`. 

I have a simple class (called `CollisionGroup`) which stores a `std::vector` of pointers to `CollisionObjects` that share a transformation (i.e. they are part of the same real-life rigid body). The only methods on `CollisionGroup` are things like `AddCollisionObject(object)` which appends the object to its list of objects, `CollisionObjects()` which returns the list of pointers, and `setTransform(const hpp::fcl::Transform3f& T)` which calls `setTransform(T)` on each of the collision objects in its list (therefore making sure they all have the same transform).

# Setup of the collision objects:

First, all of the collision-models for each sub-mesh are created:
```cpp
using CollisionManager = hpp::fcl::DynamicAABBTreeCollisionManager;
using Model = hpp::fcl::Convex<hpp::fcl::Triangle>;

// Create each real-life collision object in a loop:
  for (const auto& real_life_object : real_life_objects_) {

    // All of the sub collision-models will be stored together in a "group". This way,
    // their transformations can be set all together. This "group" isn't much more than
    // a std::vector storing all of the shared_ptr s
    CollisionGroup collision_group;

    // vertices and triangles to create each sub-collision object:
    std::vector<std::vector<Eigen::Vector3d>>    object_vertices{real_life_object.n_submodels(), std::vector<Eigen::Vector3d>()};
    std::vector<std::vector<hpp::fcl::Triangle>> object_triangles{real_life_object.n_submodels(), std::vector<hpp::fcl::Triangle>()};

    for (int i_submodel = 0; i_submodel < real_life_object.n_submodels(); ++i_submodel) {

      // Here there is some code to load the mesh file of this "sub-collision-object" ...
      // vertices are loaded into object_vertices[i_submodel], triangles are loaded into object_triangles[i_submodel] ...

      // Create the collision geometry for this portion of the object:
      auto collision_sub_geometry = std::make_shared<Model>(
          false,
          object_vertices[i_submodel].data(),
          object_vertices[i_submodel].size(),
          object_triangles[i_submodel].data(),
          object_triangles[i_submodel].size());

      // Compute the local AABB:
      collision_sub_geometry->computeLocalAABB();

      // Create the collision object based on the geometries and identity pose:
      collision_group.AddCollisionObject(new hpp::fcl::CollisionObject(collision_sub_geometry,  hpp::fcl::Transform3f::Identity()));

      // Compute the AABB for the new collision object
      collision_group.CollisionObjects().back()->computeAABB();
    }

    // Add the new collision group to the internal list of collision groups:
    collision_groups_.emplace_back(collision_group);
  }

```

Next, all of the collision objects are registered to a single collision manager - this manager manages all of the objects that move. There is a separate collision manager for stationary objects only (the environment). I do not care about collisions between moving objects, only between moving objects and the environment.

```cpp
  // Create a collision manager:
  moving_objects_manager_ = new CollisionManager();

  // Register all of the collision objects with the collision manager:
  for (auto& collision_group : collision_groups_) {
    moving_objects_manager_->registerObjects(collision_group.CollisionObjects());
  }

  // Initialize the manager:
  moving_objects_manager_->setup();
```

# Usage of the collision objects:
Basically, I have a list of transformations, one for each `CollisionGroup`. 

In my `CollisionGroup` class, I have a method called `setTransform(const hpp::fcl::Transform3f& T)`, which simply calls the `setTransform` of each collision object in the group, and gives them all the same transform `T`.

I then run the usual `collision_object->computeAABB();` on every individual object after the transformations are set. Finally, I call `moving_objects_manager_->update()` to update the collision manager after setting the transformations.

I then use the collision managers as I was previously for `BVHModel`s, like this:

```cpp
hpp::fcl::CollisionCallBackDefault callback;
moving_objects_manager_->collide(environment_collision_manager_, &callback);
```

# Things I have tested so far
- The transformations of each `CollisionObject` seem to be set correctly - I can tell by checking the `getTransform()` method on each collision object.
- The meshes themselves are correct (they haven't changed from before, and I've at least confirmed that things like the `computeVolume` give me the results I would expect.
- The `CollisionGroup`s definitely contain all of the correct collision object pointers.

# Other bugs I am getting:
If I run enough random configurations, I eventually inevitably get the following error:
```cpp
fcl/src/narrowphase/gjk.cpp:328: void hpp::fcl::details::getSupportFuncTpl(const hpp::fcl::details::MinkowskiDiff&, const Vec3f&, bool, hpp::fcl::Vec3f&, hpp::fcl::Vec3f&, hpp::fcl::support_func_guess_t&, hpp::fcl::details::MinkowskiDiff::ShapeData*) [with Shape0 = hpp::fcl::details::LargeConvex; Shape1 = hpp::fcl::details::SmallConvex; bool TransformIsIdentity = false; hpp::fcl::Vec3f = Eigen::Matrix<double, 3, 1>; hpp::fcl::support_func_guess_t = Eigen::Matrix<int, 2, 1>]: Assertion `!NeedNormalizedDir || dirIsNormalized || fabs(dir.normalized().squaredNorm() - 1) < 1e-6' failed.
Aborted (core dumped)
```

I guess the GJK algorithm uses normalized search directions? Anyways, I am not sure if this error has something to do with the setup of my objects too?

Thanks again for all your help!

**Comment by jcarpent** [comment]
Are you sure your geometry is really convex ?

**Comment by jcarpent** [comment]
You can use the meshLoader to load directly your mesh as a BVHModel and then transform it as a convex one using QHull.

**Comment by JeffHough** [comment]
The meshes are definitely convex in the usual sense, although I'm not sure if the `Convex` type has some particular setup which makes the meshes not valid.

I just tried your suggestion (setting up each collsion object as a `BVHModel` and then using `buildConvexRepresentation` and adding the geometries into the collision objects as `collision_geometry->convex` instead of just `collision_geometry`. The results are still not exactly the same as they were before, but things are much, much closer now!

**Comment by JeffHough** [comment]
@jcarpent interesting, the `security_margin` in the `hpp::fcl::CollisionCallBackDefault` no longer seems to work when the meshes are convex?

**Comment by jcarpent** [comment]
Could you give a precise example to reproduce your issue? Hard to follow on our side otherwise.

**Comment by JeffHough** [comment]
Sure. It has to be somewhat simplified since I cannot show code that belongs to the company I work for.

But in my unit tests, I perform the following:

```
  auto distance_data0 = CollisionDetector()->Distance(/* some pose data */);
  bool correct_distance0 = distance_data0.min_distance < 0.09 && distance_data0.min_distance > 0.08;
  EXPECT_TRUE(correct_distance0);

```

which basically checks that the minimum distance is between 8 and 9 cm, which is what I would expect. This check passes (as do all of the other distance checks).

Then, I perform collision checks using security margins:
```
auto collision_data0 = CollisionDetector()->Collide(/*same poses as before*/, 0.08);
EXPECT_FALSE(collision_data0.isCollision());

auto collision_data1 = CollisionDetector()->Collide(/*same poses as before*/, 0.09);
EXPECT_TRUE(collision_data1.isCollision());
```

And that `Collide` function looks like this:

```
hpp::fcl::CollisionResult CollisionDetector<N>::Collide(/* poses */, const float& security_margin) {
  
  // There is some code to update the poses of each object.

  // Create the callback function:
  hpp::fcl::CollisionCallBackDefault cb;

  // set the security margin
  cb.data.request.security_margin = security_margin;

  // check for collision in new pose:
  moving_objects_manager_->collide(environment_objects_manager_, &cb);

  return cb.data.result;
}
```

So in other words, what my test does is sets a security margin of 8cm and checks for collision (I know from the distance test that this should _not_ result in collision, since the distance is greater than 8cm). I then set the security margin to 9cm (this _should_ result in collision, since I know from the distance test that the distance is _less_ than 9cm).

If I am using the `geometry->convex` for the colllision object geometries, then **all** of my checks on using `security_margin` fail, regardless of how high I set the margin. If I am just using the `BVHModel`, then all of my checks pass, which is the expected behaviour.

**Comment by JeffHough** [comment]
Okay - I am guessing there is not currently a standard way of doing that? I could edit the `computeAABB()` function to include some security_margin as well. 

Are there any broadphase managers where the security margin wouldn't be an issue?

**Comment by jcarpent** [comment]
I think they all use this information. Security margin is something we have added in top of the origin FCL.
It exploits GJK to stop if we may garanty that the security margin is fulfilled.

Yet, the BroadPhases have been reimported mostly without accounting for this feature
The only way to retrieve is by manullay doing first computeAABB() of a given CollisionGeometry, then expanding it with the positive security margin.
In Pinocchio, we use a CollisionRequest per CollisionPair, which allows for being more flexible. This explains why we cannot use a single security margin.
We can try to add in the future a way to easily account for different broadphase checkers too.

**Comment by JeffHough** [comment]
Actually, it looks like there are two overloads of `AABB::overlap`, one of which includes a collision request (which then passes the `security_margin`), i.e., in the `AABB.cpp` function

```
bool AABB::overlap(const AABB& other, const CollisionRequest& request,
                   FCL_REAL& sqrDistLowerBound) const {
  const FCL_REAL break_distance_squared =
      request.break_distance * request.break_distance;

  sqrDistLowerBound =
      (min_ - other.max_ - Vec3f::Constant(request.security_margin))
          .array()
          .max(0)
          .matrix()
          .squaredNorm();
  if (sqrDistLowerBound > break_distance_squared) return false;

  sqrDistLowerBound =
      (other.min_ - max_ - Vec3f::Constant(request.security_margin))
          .array()
          .max(0)
          .matrix()
          .squaredNorm();
  if (sqrDistLowerBound > break_distance_squared) return false;

  return true;
}
```

However, the broadphase managers do not call this version. I.e., in `broadphase_dynamic_AABB_tree.cpp` line 99:
`if (obb1.overlap(obb2))  ... `

Might it work to simply change these calls in the broadphase managers to the overload of `AABB::overlap` which includes the collision request (and by extension, includes the security margin)?

**Comment by jcarpent** [comment]
Yes, it shoule be one way, but this is not generic enough as we can decide to have a CollisionRequest per CollisionPair ...

**Comment by jcarpent** [comment]
I think the best way would be to add in the `CollisionCallBack` a method for checking the collision between BV volumes, which could account for the security margin ;)

**Comment by jcarpent** [comment]
a working solution is simply growing the AABB volumes ;)

**Comment by JeffHough** [comment]
That sounds good - in terms of manually growing the AABB volumes, what would you recommend is the best way to do that without editing the hpp-fcl codebase?

I can see that the `CollisionObject` has methods to `getAABB` and `computeAABB`, but I don't see anything like `setAABB` where I could do something like a `setAABB(getAABB() + security_margins)`?

**Comment by jcarpent** [comment]
`getAABB().inflate(security_margin/2)` should work

**Comment by JeffHough** [comment]
Okay, that is working for me now! Although I do not use `security_margin/2`, but `security_margin` on its own, and I think the function is called `expand`!

**Comment by JeffHough** [comment]
Thanks so much for your help!

**Comment by jcarpent** [comment]
I'm trying to do that from memory. So sorry for the miss-naming ;)
